persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/LogAppender.java

```diff
@@ -1,124 +1,207 @@
-               temporaryTable.set(request.getKey(), logFile.fileId, offset);
-               IndexRequest indexRequest = IndexRequest.update(request.getKey(), raw(request.getSerializedKey()),
-                     logFile.fileId, offset, request.length());
-               request.setIndexRequest(indexRequest);
-               indexQueue.put(indexRequest);
-               currentOffset += request.length();
-            } else {
-               if (syncWrites) {
-                  logFile.fileChannel.force(false);
-               }
-               queue.notifyAndWait();
-            }
+         if (logFile == null) {
+            logFile = fileProvider.getFileForLog();
+            log.debug("Appending records to " + logFile.fileId);
+         }
+         if (request.isClear()) {
+            logFile.close();
+            currentOffset = 0;
+            logFile = null;
+            // TODO: Technically other writes won't be notified if sync write is enabled until
+            // the timeout above
+            delay = AsyncProcessor.create();
+            completeRequest(request);
+            return;
+         } else if (request.isPause()) {
+            delay = AsyncProcessor.create();
+            completeRequest(request);
+            return;
+         } else if (request.isResume()) {
+            delay.onComplete();
+            completeRequest(request);
+            return;
+         }
+         if (currentOffset != 0 && currentOffset + request.length() > maxFileSize) {
+            // switch to next file
+            logFile.close();
+            compactor.completeFile(logFile.fileId);
+            logFile = fileProvider.getFileForLog();
+            currentOffset = 0;
+            log.debug("Appending records to " + logFile.fileId);
          }
+         long seqId = nextSeqId();
+         EntryRecord.writeEntry(logFile.fileChannel, request.getSerializedKey(), request.getSerializedMetadata(),
+               request.getSerializedInternalMetadata(),
+               request.getSerializedValue(), seqId, request.getExpiration(), request.getCreated(), request.getLastUsed());
+         int offset = request.getSerializedValue() == null ? ~currentOffset : currentOffset;
+         temporaryTable.set(request.getKey(), logFile.fileId, offset);
+         IndexRequest indexRequest = IndexRequest.update(request.getKey(), raw(request.getSerializedKey()),
+               logFile.fileId, offset, request.length());
+         request.setIndexRequest(indexRequest);
+         // TODO: This technically can block, but that should be okay.. or maybe rewrite index part as well?
+         indexQueue.put(indexRequest);
+         if (!syncWrites) {
+            completeRequest(request);
+         }
+         currentOffset += request.length();
       } catch (Exception e) {
-         queue.notifyError();
-         throw new RuntimeException(e);
+         request.completeExceptionally(e);
```

Comment:
Is it worth adding a debug log here?

Reply:
We could, it shouldn't really ever happen. I can add it.